File: tests/python/test_remesh.py
import os
from itertools import product

import pytest

import tinymesh as tms

CWD = os.getcwd()

model_dir = 'data/models'
filenames = [
    'torus.obj',
    'fandisk.ply',
    'bunny_mini.ply',
]

methods = [
    tms.remesh_triangular,
    lambda m: tms.simplify_qem(m,
                               m.num_faces() // 2),
    lambda m: tms.simplify_qem(m,
                               m.num_faces() // 10),
]


@pytest.mark.parametrize("filename", filenames)
def test_hole_fill(filename):
    filename = os.path.join(CWD, model_dir, filename)
    mesh = tms.Mesh(filename)

    try:
        mesh.fill_holes()
    except Exception:
        pytest.fail('Failed!')

    assert mesh.verify(), "Mesh verification failed!"


@pytest.mark.parametrize("method, filename", product(methods, filenames))
def test_remesh_method(method, filename):
    filename = os.path.join(CWD, model_dir, filename)
    mesh = tms.Mesh(filename)
    mesh.fill_holes()

    try:
        method(mesh)
    except Exception:
        pytest.fail('Failed!')

    assert mesh.verify(), "Mesh verification failed!"

File: tests/test_remesh.cpp
#include "gtest/gtest.h"

#include <cstdio>
#include <sstream>
#include <vector>
#include <tuple>
#include <algorithm>

#include "tinymesh/tinymesh.h"
using namespace tinymesh;

#include "test_utils.h"

class MeshRemeshTest : public TinyMeshBaseTestWithParam {
public:
    MeshRemeshTest() = default;
    virtual ~MeshRemeshTest() = default;

    void SetUp() override {
        mesh.load(filename);
    }
};

TEST_P(MeshRemeshTest, HoleFill) {
    EXPECT_NO_FATAL_FAILURE(mesh.fillHoles());
    EXPECT_TRUE(mesh.verify());
}

TEST_P(MeshRemeshTest, RemeshTriangular) {
    EXPECT_NO_FATAL_FAILURE(mesh.fillHoles());
    EXPECT_NO_FATAL_FAILURE(remeshTriangular(mesh));
}

TEST_P(MeshRemeshTest, SimplifyQEMOneHalf) {
    EXPECT_NO_FATAL_FAILURE(mesh.fillHoles());
    EXPECT_NO_FATAL_FAILURE(simplifyQEM(mesh, (int)mesh.numFaces() / 2));
}

TEST_P(MeshRemeshTest, SimplifyQEMOneTenth) {
    EXPECT_NO_FATAL_FAILURE(mesh.fillHoles());
    EXPECT_NO_FATAL_FAILURE(simplifyQEM(mesh, (int)mesh.numFaces() / 10));
}

static std::vector<std::string> filenames = {
    "box.obj",
    "torus.obj",
    "fandisk.ply",
    "bunny_mini.ply",
};

INSTANTIATE_TEST_SUITE_P(, MeshRemeshTest, ::testing::ValuesIn(filenames));
